### ISSUE TITLE
Check for coherence after any (3-ways) merge

### ISSUE THREAD

**Issue by Phillipus** [archi-contribs/archi-modelrepository-plugin#7]
We need to handle the merge conflict case of User A deleting elements from the model, while User B has pushed changes that still reference those elements in a diagram.

**Comment by Phillipus** [comment]
User A commits changes to a Diagram and pushes
User B deletes elements from the Diagram and the Model and commits and pulls

There is now a conflict.

For User B only conflicting files will be shown. The deleted element files are not conflicted and are removed from the repo and not shown in the conflict dialog. However, if User B selects "theirs" for the Diagram file in the merge conflict dialog and commits the merge then there is a problem because the diagram is now referencing the deleted elements.

**Comment by Phillipus** [comment]
For reference:

https://github.com/archi-contribs/archi-grafico-plugin/issues/3

**Comment by Phillipus** [comment]
From @jbsarrodie :

When importing, grafico will log this as an error, but for the moment there is no automatic management of this because this requires user inputs.

An idea : if the pull fails because of conflicts, then we ask user to decide which version of conflicting elements to use (that's what we have now). But then instead of commiting we should : 

- Either do the commit and reload model through grafico and warn user of the issue and remove diagram elements that reference deleted concepts
- Or reload the model and warn user of the issue and then revert to previous state before merge attempt.

**Comment by jbsarrodie** [comment]
I'll discuss this tomorrow morning to have feedback from potential users.

**Comment by jbsarrodie** [comment]
After having discussed about this topic, it seems now clear that the best thing to do is to check model for incoherence just **before** the merge commit, and even when no conflicts have been detected (as incoherence can happen in very simple case with not git conflict: user A remove element E while user B uses E in a new view).

So we should :
* stop the merge process just before the commit (similar to `git pull --no-commit`).
* if no real merge (fast-forward) then commit and push
* user solves conflicts (if any)
* try to load the model folder and check for incoherences.
* if no incoherences found, then commit and push
* if any then ask the user what to do (a simple, automated, solution would be to delete "things" that reference deleted concepts).
* if user choose "manual resolution" then don't commit and let the user fix the mess (two alternatives: roll-back as before merge attempt, or let things as they were after the merge attempt)
* user fixes the issues, commit and publish

**Comment by jbsarrodie** [comment]
Renaming this issue as this impact also merge without conflicts

**Comment by Phillipus** [comment]
Can you move it to Sprint 2?

**Comment by jbsarrodie** [comment]
I move it back to ToDo for sprint #1 knowing that it could be postponed.

**Comment by jbsarrodie** [comment]
Some thoughts...

When inconsistency is detected, we should provide two (global) options:
- Delete broken "things" (relationships, diagram elements...) that reference missing concepts
- Keep broken "things" and restore missing concepts

The last point was not straightforward, but I think we have a  way of doing it: use Git to restore them. This can be done because Grafico importer will provide us the list of ID of missing concepts, and XML filenames are generated from ID, so given an _ID_ we know that the filename will be *__ID_.xml

**Comment by Phillipus** [comment]
OK.

> use Git to restore them

How? From the commit that had it last?

**Comment by jbsarrodie** [comment]
>How? From the commit that had it last?

As we are merging two commits (master and origin/master), we have to restore it from one of those. By design a missing concept will exist in only one of them because it has been removed from the other one.

My only concern here is how to do it efficiently. I can imagine that we first build the list of missing concepts (in fact get if from grafico importer) and then we walk through both commit trees, checking for each entry if it matches a missing concepts, and if yes restore it. This way we only walk through each tree only once (and not once per missing concept).

Warning: I write here missing **concept** but it can **also be a view** as we can reference a view in another view, so this has to be "open and generic"

**Comment by Phillipus** [comment]
OK, experiments are in progress...

**Comment by Phillipus** [comment]
New branch created "resolve_merge"

> Delete broken "things" (relationships, diagram elements...) that reference missing concepts

- Added a new GraficoResolutionHandler class.
- The GraficoModelImporter adds problems to this class.
- The first implementation deletes problem objects.
- It works for simple cases
- Doesn't work when connections and relations are deleted

Needs more investigation!

**Comment by Phillipus** [comment]
This is not so easy.

It depend on what the user selects in the merge conflict dialog. If they select "theirs" it may be that a number of concepts will have to be re-added back again *and* some deleted.

**Comment by jbsarrodie** [comment]
This is not linked to conflict: this can happen even without any conflict. So this should be a "standalone" step.

In case of conflicts, this step has to be done after conflicts resolution.

**Comment by Phillipus** [comment]
I know, but the user choice will affect the "standalone" step. So, if the user opts to keep the online version, the process of deleting missing things is not straightforward. We need to discuss this deeper.

**Comment by Phillipus** [comment]
In branch "resolve_merge" the Refresh action has code to check this.

**Comment by jbsarrodie** [comment]
>I know, but the user choice will affect the "standalone" step. So, if the user opts to keep the online version, the process of deleting missing things is not straightforward.

Er... I don't understand. It really doesn't depend on how user got to the current model version, if it is inconsistent it has to be fixed. At this stage It really no more depends on what the user did before. So if the auto-fix is "delete dangling things", then we delete them. If the choice is "restore missing things", then we restore them.

>We need to discuss this deeper.

Seems so ;-)

**Comment by Phillipus** [comment]
I have a case:

- User A adds a new connection (+ relationship) and commits and pushes to the repo
- Meanwhile, User B in his local copy deletes some concepts (including the ones that User A added a connection to) and commits this change.
- User B does a "Refresh". Git "Pull" is done and a conflict dialog comes up. User B chooses "theirs". Theirs does not have the deleted concepts in the diagram that B made.
- After this the GraficoResolutionHandler then deletes all missing things but the new connection is still there but the associated nodes are missing.

So, User B opts to re-instate the things that he had previously deleted by choosing theirs.

**Comment by Phillipus** [comment]
Note - I've rebased the  "resolve_merge" branch to include latest fixes.

**Comment by jbsarrodie** [comment]
>After this the GraficoResolutionHandler then deletes all missing things but the new connection is still there but the associated nodes are missing.

How the new connection could still be there ? The connection is related to a relationship that will be dangling so when removing the relationship the connection will be removed too.

>So, User B opts to re-instate the things that he had previously deleted by choosing theirs.

I don't think that's how a user will perceive it: when solving a conflict on a view, user will focus on the diagram, so if choosing a view means that some deletions have to be "rolled-back" then that's the indended behavior as this will allows the diagram to be ok.

And again, in several (most?) cases inconsistencies will appear without a conflict (me and my team had one case last friday).

**Comment by jbsarrodie** [comment]
I had a look at your "deleteProblemObjects" method and it seems to work fine.

I've started to think about the other option (let's call it "restoreProblemObjects") and I think this involves:
* have access to the git folder (not done for the moment as only AbstractModelAction has, but not GraficoResolutionHandler)
* Walk through commit trees that preceed the merge (HEAD^1 and HEAD^2), checking for each entry if it matches a missing concepts, and if yes restore it

Your idea ?

**Comment by Phillipus** [comment]
I've commented out the deleteProblemObjects() call in Refresh Action for now, because I'm unsure of its impact alone. Also, GraficoResolutionHandler is now a protected helper class for GraficoModelImporter.

There are going to be problems either with or without merge conflicts. In my test case I've found that deleteProblemObjects() is not enough on its own. For example, If a user chooses "theirs" for a view xml file in the merge dialog, then it's likely that some objects will need to be restored as well. So it's not a case of deleting problem objects OR re-instating them but sometimes both.

**Comment by jbsarrodie** [comment]
>There are going to be problems either with or without merge conflicts. In my test case I've found that deleteProblemObjects() is not enough on its own. For example, If a user chooses "theirs" for a view xml file in the merge dialog, then it's likely that some objects will need to be restored as well. So it's not a case of deleting problem objects OR re-instating them but sometimes both.

Can you provide me a use-case to reproduce it (or just prepare it on bitbucket if it helps) ?

**Comment by Phillipus** [comment]
Here's a use case, but it's better you set it up so that you can synchronise the instances:

On instance 1 (I use a laptop)

1. make sure everything is up to date
2. add a connection between two nodes in a diagram (this will update the ArchimateDiagramModel_xxxxxx.xml file and add a Relationship.xml file)
3. push to server

On instance 2 (I use a desktop)

1. don't Refresh just yet
2. delete some nodes from the same diagram (including a node where the connection was added in the other instance)
3. commit
4. now Refresh and go through the merge dialog selecting "theirs"


Note - you can uncomment line 149 in AbstractModelAction

**Comment by jbsarrodie** [comment]
Hi,

I have tested the new implementation which restore missing concepts. It works really well (at least for simple cases I tested).

After several discussions, I confirm that (from a user point of view) it almost never makes sens to delete dangling elements, so I suggest that in all cases we restore missing concepts but provide feedback to the user after doing so. This can be a dialog that explains that some concepts have been restored because they were needed, and then provide the list of those concepts (with name and type).

An enhanced version (from a UX point of view) would list only concepts that the user have deleted but have been restored (not the one that have been deleted by someone else and restored).

**Comment by jbsarrodie** [comment]
I've tested this new version and behavior is good with one exception: a commit is done for the merge and concepts are being restored but not commited.

It would be better to do the merge in the tree without a commit, and then commit once **after** restoration (if any). This would make it virtually impossible for a user to undo a commit and get a screwed model.

**Comment by Phillipus** [comment]
Which action is this? Refresh, or Publish?

**Comment by jbsarrodie** [comment]
It is needed for any 3 ways merge, so this should be done in both cases.

**Comment by Phillipus** [comment]
Yes, Publish action needs this now.

**Comment by Phillipus** [comment]
> It would be better to do the merge in the tree without a commit

Not sure this is possible, since it is the Pull action that automatically does the merge. And a merge needs to occur so that the Grafico Import can report the missing concepts.

**Comment by Phillipus** [comment]
> It would be better to do the merge in the tree without a commit

Wait. I think I see the problem. Publish is doing merge and commit. Refresh is just doing merge. Fixing... ;-)

**Comment by jbsarrodie** [comment]
I've juste tested latest version of code (the one tagged 0.2.4)... Still not good with different behavior between Refresh and Push....

Here is step to reproduce issue...

On a first instance of Archi (user 1):
* create a new, empty model and add it to collaboration workspace
* add one concept to the model (not used in view)
* save, commit, publish

On a second instance of Archi (user 2):
* import the model to the collaboration workspace
* delete the concept
* save, commit, publish

Back on first instance of Archi:
* use the concept in a view
* save and commit

Now we have to possible actions (Refresh or Publish) with different (but both unwanted) behavior...

If you **publish**, you get a warning telling you that the concept has been restored and then the action finishes. But what you get is a merge commit ("Merge branch 'master' of ....") that has been created and pushed, and an unsaved/uncommitted change (restoration of the concept).

If you **refresh**, you get a warning telling you that the concept has been restored and then the action finishes. But what you get is a merge commit ("Merge branch 'master' of ....") that has been created, followed by another commit that contains the restored concept.

The intended behavior is:

If you **publish**, you get a warning telling you that the concept has been restored and then the action finishes. Then what you get is a merge commit ("Merge branch 'master' of ....") that includes the restored concept, and that has been pushed.

If you **refresh**, you get a warning telling you that the concept has been restored and then the action finishes. Then what you get is a merge commit ("Merge branch 'master' of ....") that includes the restored concept.

So this requires the equivalent of a "git merge --no-commit".

Is it clearer ?

**Comment by Phillipus** [comment]
The first commit ("Merge branch 'master' of ....") is done automatically by the Pull action. So I guess we need to do something else. Not sure what yet...

**Comment by jbsarrodie** [comment]
on command line you can do a `git pull --no-commit`, so there must be an equivalent with JGit.

**Comment by Phillipus** [comment]
> on command line you can do a git pull --no-commit, so there must be an equivalent with JGit.

Sadly, no... ;-)

**Comment by jbsarrodie** [comment]
Mmm just looked at the JGit API, and I don't see the equivalent. Maybe we can let the pull do a commit, but do a `commit --amend` after (and this option exists in JGit, I've checked)

**Comment by jbsarrodie** [comment]
BTW we coudl also add the list of restored concepts in the commit message...

**Comment by Phillipus** [comment]
I'll take a look at this and see what could be done.

**Comment by Phillipus** [comment]
I've done some major refactoring:

- Refresh and Publish actions should be more consistent now
- One commit is done (uses commit --amend) but the commit message is not the previous one
- Commit message includes the list of restored concepts

Now I need to solve the EOL issue...

**Comment by jbsarrodie** [comment]
I've just tested and this is exactly the intended behavior. Perfect !
